napari/plugins/_npe2.py

```diff
@@ -490,29 +509,196 @@ def _add_sample(
     return submenu, sample_actions
 
 
+def _get_widget_viewer_param(widget_callable, widget_name) -> str:
+    """Get widget parameter name (if any) and check type."""
+    from napari.viewer import Viewer
+
+    if inspect.isclass(widget_callable) and issubclass(
+        widget_callable,
+        (QWidget, Widget),
+    ):
+        widget_param = ""
+        # Inspection can fail when adding to bundle as it thinks widget is
+        # a builtin
+        try:
+            sig = inspect.signature(widget_callable.__init__)
+        except ValueError:
+            pass
+        else:
+            for param in sig.parameters.values():
+                if param.name == 'napari_viewer':
+                    widget_param = param.name
+                    break
+                if param.annotation in (
+                    'napari.viewer.Viewer',
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    elif isinstance(widget_callable, MagicFactory) or inspect.isfunction(
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "'{widget}' must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or function. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+            )
+        )
+    return widget_param
+
+
+def _widget_callback_implementation(
```

Comment:
```suggestion
def _get_widget_callback(
```

I think it should be more explicit that this is returning a callback. Maybe a return type annotation would be nice also